From ff3abfdaeed4bd5b20aee611b094169a941b69a1 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 7 Dec 2021 14:57:41 +0100
Subject: [PATCH 0162/1695] clk: add CLK_OPS_PARENT_ENABLE flags

During a change parent clock, we have to enable the old parent and the
new parent clocks if CLK_OPS_PARENT_ENABLE flag is set.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c568c9e8237cf24bb5ff2ee4173a7a69b1dc160b clk: add CLK_OPS_PARENT_ENABLE flags]
Change-Id: I389dc0739283b88840fa70bcdad9cba154e3374b
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
---
 core/drivers/clk/clk.c     | 26 ++++++++++++++++++++------
 core/include/drivers/clk.h |  1 +
 2 files changed, 21 insertions(+), 6 deletions(-)

diff --git a/core/drivers/clk/clk.c b/core/drivers/clk/clk.c
index 51a1814d9..3617f18ed 100644
--- a/core/drivers/clk/clk.c
+++ b/core/drivers/clk/clk.c
@@ -273,6 +273,7 @@ static TEE_Result clk_get_parent_idx(struct clk *clk, struct clk *parent,
 static TEE_Result clk_set_parent_no_lock(struct clk *clk, struct clk *parent,
 					 size_t pidx)
 {
+	struct clk *old_parent = clk->parent;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	bool was_enabled = false;
 
@@ -281,9 +282,16 @@ static TEE_Result clk_set_parent_no_lock(struct clk *clk, struct clk *parent,
 		return TEE_SUCCESS;
 
 	was_enabled = clk_is_enabled_no_lock(clk);
-	/* Call is needed to decrement refcount on current parent tree */
-	if (was_enabled)
-		clk_disable_no_lock(clk);
+
+	if (clk->flags & CLK_OPS_PARENT_ENABLE) {
+		res = clk_enable_no_lock(parent);
+		if (res)
+			panic("Failed to re-enable clock after setting parent");
+
+		res = clk_enable_no_lock(old_parent);
+		if (res)
+			panic("Failed to re-enable clock after setting parent");
+	}
 
 	res = clk->ops->set_parent(clk, pidx);
 	if (res)
@@ -294,12 +302,18 @@ static TEE_Result clk_set_parent_no_lock(struct clk *clk, struct clk *parent,
 	/* The parent changed and the rate might also have changed */
 	clk_compute_rate_no_lock(clk);
 
-out:
-	/* Call is needed to increment refcount on the new parent tree */
+	/* Call is needed to decrement refcount on current parent tree */
 	if (was_enabled) {
-		res = clk_enable_no_lock(clk);
+		res = clk_enable_no_lock(parent);
 		if (res)
 			panic("Failed to re-enable clock after setting parent");
+
+		clk_disable_no_lock(old_parent);
+	}
+out:
+	if (clk->flags & CLK_OPS_PARENT_ENABLE) {
+		clk_disable_no_lock(old_parent);
+		clk_disable_no_lock(parent);
 	}
 
 	return res;
diff --git a/core/include/drivers/clk.h b/core/include/drivers/clk.h
index a6ce2484b..154b3b061 100644
--- a/core/include/drivers/clk.h
+++ b/core/include/drivers/clk.h
@@ -14,6 +14,7 @@
 /* Flags for clock */
 #define CLK_SET_RATE_GATE	BIT(0) /* must be gated across rate change */
 #define CLK_SET_PARENT_GATE	BIT(1) /* must be gated across re-parent */
+#define CLK_OPS_PARENT_ENABLE	BIT(2) /* parent need enable during re-parent */
 
 /**
  * struct clk - Clock structure
-- 
2.47.2

